desc/protoparse/parser.go

```diff
@@ -868,7 +868,10 @@ func (r *parseResult) asMethodDescriptor(node *methodNode) *dpb.MethodDescriptor
 	if node.output.streamKeyword != nil {
 		md.ServerStreaming = proto.Bool(true)
 	}
-	if len(node.options) > 0 {
+	// protoc always adds a MethodOptions if there are brackets
+	// We do the same to match protoc as closely as possible
+	// https://github.com/protocolbuffers/protobuf/blob/0c3f43a6190b77f1f68b7425d1b7e1a8257a8d0c/src/google/protobuf/compiler/parser.cc#L2152
+	if len(node.options) > 0 || node.brackets {
```

Comment:
How about we instead detect this via `node.options != nil`, and in the grammar, for the `rpcOptions` and `rpcOption` productions, use `[]*optionNode{}` as the empty sentinel instead of `nil`.

Reply:
I tried that, but that doesn't work - see https://github.com/protocolbuffers/protobuf/blob/0c3f43a6190b77f1f68b7425d1b7e1a8257a8d0c/src/google/protobuf/compiler/parser.cc#L2152 - you only generate this if there are brackets. If you do the empty sentinel, it's the same as removing the `if len(node.options) > 0`, and then `desc/protoparse` generates MethodOptions for methods that `protoc` does not. This is the only way to do it - `protoc` strictly switches on brackets.

Reply:
> I tried that

Including the grammar changes I suggested? What I was suggesting should only include it if there are brackets since it would change the grammar to produce a non-nil `options` if there are brackets (but it would remain nil if no brackets).

Reply:
Ah I see what you're saying. Let me try that.

Reply:
Yep, that worked! Updated.